crates/bevy_ecs/src/query/state.rs

```diff
@@ -949,34 +882,17 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
 
                     let mut offset = 0;
                     while offset < table.entity_count() {
-                        let func = func.clone();
+                        let mut func = func.clone();
                         let len = batch_size.min(table.entity_count() - offset);
                         let task = async move {
-                            let mut fetch = Q::init_fetch(
-                                world,
-                                &self.fetch_state,
-                                last_change_tick,
-                                change_tick,
-                            );
-                            let mut filter = F::init_fetch(
-                                world,
-                                &self.filter_state,
-                                last_change_tick,
-                                change_tick,
-                            );
-                            let tables = &world.storages().tables;
-                            let table = tables.get(*table_id).debug_checked_unwrap();
-                            let entities = table.entities();
-                            Q::set_table(&mut fetch, &self.fetch_state, table);
-                            F::set_table(&mut filter, &self.filter_state, table);
-                            for row in offset..offset + len {
-                                let entity = entities.get_unchecked(row);
-                                let row = TableRow::new(row);
-                                if !F::filter_fetch(&mut filter, *entity, row) {
-                                    continue;
-                                }
-                                func(Q::fetch(&mut fetch, *entity, row));
-                            }
+                            let table = &world
+                                .storages()
+                                .tables
+                                .get(*table_id)
+                                .debug_checked_unwrap();
+                            let batch = offset..offset + len;
+                            self.iter_unchecked_manual(world, last_change_tick, change_tick)
```

Comment:
not particularly thrilled with us creating a `QueryIter` that isnt actually allowed to access all the component data it should have access to. I've always interpreted `iter_unchecked_manual`'s safety invariant to be that it must be legal for the iterator to access all matched components which is absolutely not the case here...

(same applies to the non-dense branch below)

I am not sure if this is actually exploitable or not (regardless I do not like this). For example when this calls `QueryState::iter_unchecked_manual` which calls `QueryIter::new` which calls `QueryIterationCursor::init` which calls `Q::init_fetch` we end up in arbitrary user code which is rather poorly documented safety wise as to what it is allowed to access from `World` but I think that the intention is that it has access to anything specified in `fn update_archetype_component_access`

Reply:
Wasn't that what we were already doing here? Would it be better if we moved the shared functions onto `QueryIterationCursor` instead and manually constructed on here for iterating over a given batch from an archetype/table?

Reply:
Well... we weren't making a `QueryIter` which is better than this. But yeah we were calling `init_fetch` before too so this PR is not dramatically worse than how things were before this PR.

Reply:
It sounds to me like `init_fetch` needs to be clarified to explicitly say that accessing component data is disallowed. All bevy implementations of that function are either empty, pass throughs (tuple and the derive), or `world.storages().sparse_sets.get(component_id).debug_checked_unwrap()`.

Honestly I can't think of a way to safely access component data at that point given we haven't even called `F::init_fetch` or subsequent functions and thus can't know what effect filters have on our component access which is certainly necessary for knowing what component data we can safely access.

You need to wait until `Q::fetch` to touch the data as any earlier you haven't called `F::filter_fetch` which is needed to properly determine `Without` and similar filters that are used to guarantee disjoint access of parallel work.

Reply:
If this doesn't dramatically worsen our current soundness stance, I think it's best to get this PR finally merged and file an issue to deal with this in a follow up PR (preferably before 0.13 ships).

Reply:
Filed this under #10752.